File: frontend/src/api/coaching.ts
import {
  getCoachingInsightsApiV1CoachingSoforIdInsightsGet,
  sendCoachingApiV1CoachingSoforIdSendPost,
  getCoachingEffectivenessApiV1CoachingEffectivenessGet,
} from "../generated/api/coaching/coaching";

export type CoachingCategory =
  | "yakit_yonetimi"
  | "guzergah_tercihi"
  | "sofor_pratigi"
  | "diger";
export type CoachingPriority = "low" | "medium" | "high";
export type CoachingSource = "llm" | "fallback";

export interface CoachingInsight {
  category: CoachingCategory;
  pattern: string;
  evidence: string[];
  suggestion: string;
  impact_score: number;
}

export interface CoachingInsightsResponse {
  sofor_id: number;
  ad_soyad: string;
  headline: string;
  priority: CoachingPriority;
  insights: CoachingInsight[];
  generated_at: string;
  source: CoachingSource;
}

export interface SendCoachingResponse {
  sent: boolean;
  delivery_id: number | null;
  channel: string;
  sent_at: string;
}

export interface CoachingEffectivenessResponse {
  window_days: number;
  total_sent: number;
  total_evaluated: number;
  improved: number;
  worsened: number;
  improve_rate: number | null;
  avg_score_delta_pct: number | null;
  caveat: string;
}

export const coachingService = {
  /**
   * Şoförün koçluk önerilerini getirir. Backend 30dk Redis cache uygular,
   * frontend de aynı pencereyle TanStack Query cache'i ayarlar.
   */
  getInsights: async (soforId: number): Promise<CoachingInsightsResponse> => {
    const data =
      await getCoachingInsightsApiV1CoachingSoforIdInsightsGet(soforId);
    return data as unknown as CoachingInsightsResponse;
  },

  /**
   * Manuel koçluk mesajını Telegram'a ilet. Backend HTML formatla +
   * html.escape() uygular; XSS koruması sağlanmıştır.
   */
  send: async (
    soforId: number,
    message: string,
    insightCategory?: CoachingCategory,
  ): Promise<SendCoachingResponse> => {
    const data = await sendCoachingApiV1CoachingSoforIdSendPost(soforId, {
      message,
      channel: "telegram",
      insight_category: insightCategory,
    });
    return data as unknown as SendCoachingResponse;
  },

  /**
   * A.5 — Son N günde gönderilmiş koçluk mesajlarının etki özeti.
   * Caveat: skor değişimi yalnız koçluğa atfedilemez.
   */
  getEffectiveness: async (
    days: number = 30,
  ): Promise<CoachingEffectivenessResponse> => {
    const data = await getCoachingEffectivenessApiV1CoachingEffectivenessGet({
      days,
    });
    return data as unknown as CoachingEffectivenessResponse;
  },
};

File: app/schemas/coaching.py
"""Feature A — Şoför Koçluk Modülü Pydantic şemaları.

Tasarım kararları (`docs/superpowers/plans/2026-05-22-feature-a-driver-coaching-mini-plan-v2.md`):
- LLM'e gönderilen prompt'ta ad-soyad/plaka/telegram_id YOK (anonim-tam).
- Response objesi ad-soyad'ı dahili sayfa render'ı için tutar (Groq'a gitmez).
- Tüm string alanlar max_length ile sınırlı (LLM hallucination bütçesi).
"""

from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

CoachingCategory = Literal[
    "yakit_yonetimi",
    "guzergah_tercihi",
    "sofor_pratigi",
    "diger",
]

CoachingPriority = Literal["low", "medium", "high"]
CoachingSource = Literal["llm", "fallback"]


class CoachingInsightItem(BaseModel):
    """LLM'in ürettiği tek koçluk önerisi."""

    category: CoachingCategory
    pattern: str = Field(
        ..., max_length=240, description="Tespit edilen davranış kalıbı"
    )
    evidence: List[str] = Field(default_factory=list, max_length=5)
    suggestion: str = Field(
        ..., max_length=480, description="Eyleme dönüştürülebilir öneri"
    )
    impact_score: float = Field(
        0.0, ge=0, le=1, description="Önerilen aksiyon etkisi (0..1)"
    )


class CoachingInsightsResponse(BaseModel):
    """`/coaching/{id}/insights` response payload'u."""

    sofor_id: int
    ad_soyad: str = Field(..., description="Yalnız UI gösterimi için; LLM'e gitmedi")
    headline: str = Field(..., max_length=200)
    priority: CoachingPriority
    insights: List[CoachingInsightItem]
    generated_at: str = Field(..., description="ISO datetime")
    source: CoachingSource

    model_config = ConfigDict(from_attributes=True)


class SendCoachingRequest(BaseModel):
    """`POST /coaching/{id}/send` body."""

    message: str = Field(..., min_length=10, max_length=1000)
    channel: Literal["telegram"] = "telegram"
    insight_category: Optional[CoachingCategory] = Field(
        None, description="A.5 telemetri için kategori etiketi"
    )


class SendCoachingResponse(BaseModel):
    sent: bool
    delivery_id: Optional[int] = None
    channel: str
    sent_at: str


class CoachingEffectivenessResponse(BaseModel):
    """`/coaching/effectiveness?days=N` response — A.5."""

    window_days: int
    total_sent: int
    total_evaluated: int
    improved: int
    worsened: int
    improve_rate: Optional[float] = Field(None, ge=0, le=1)
    avg_score_delta_pct: Optional[float] = None
    caveat: str = Field(
        ...,
        description="UI'da açıkça gösterilecek istatistiksel uyarı",
    )
